src/libs/sfdk/dockercontainer.cpp

```diff
@@ -0,0 +1,361 @@
+{
+    Q_Q(DockerContainer);
+    Q_ASSERT(context);
+    Q_ASSERT(functor);
+
+    const QPointer<const QObject> context_{context};
+
+    auto enqueue = [=](const QStringList &arguments, CommandQueue::BatchId batch) {
+           auto runner = std::make_unique<ProcessRunner>(DockerManager::dockerPath(), arguments);
+           QObject::connect(runner.get(), &ProcessRunner::failure, Sdk::instance(), [=]() {
+               SdkPrivate::commandQueue()->cancelBatch(batch);
+               callIf(context_, functor, false);
+           });
+           ProcessRunner *runner_ = runner.get();
+           SdkPrivate::commandQueue()->enqueue(std::move(runner));
+           return runner_;
+       };
+
+    const CommandQueue::BatchId batch = SdkPrivate::commandQueue()->beginBatch();
+
+    QStringList stopArguments;
+    stopArguments.append(STOP);
+    stopArguments.append(q->m_containerId);
+
+    enqueue(stopArguments, batch);
+
+    QStringList commitArguments;
+    commitArguments.append(COMMIT);
+    commitArguments.append(q->m_containerId);
+    commitArguments.append(q->name());
+
+    enqueue(commitArguments, batch);
+
+    SdkPrivate::commandQueue()->enqueueCheckPoint(context, [=]() { functor(true); });
+    SdkPrivate::commandQueue()->endBatch();
+}
+
+void DockerContainerPrivate::probe(const QObject *context,
+        const Functor<VirtualMachinePrivate::BasicState, bool> &functor) const
+{
+    Q_Q(const DockerContainer);
+    Q_ASSERT(context);
+    Q_ASSERT(functor);
+
+    DockerManager::fetchRegisteredVirtualMachines(q, [=](const QStringList &registeredVms,
+                                                         bool ok) {
+        VirtualMachinePrivate::BasicState state;
+        if (!ok) {
+            functor(state, false);
+        } else if (!registeredVms.contains(q->name())) {
+            functor(state, true);
+        } else {
+            QStringList arguments;
+            arguments.append(CONTAINER_LIST);
+            arguments.append(QString(FILTER_PREFIX) + q->name());
+            arguments.append(FORMAT);
+            auto runner = std::make_unique<ProcessRunner>(DockerManager::dockerPath(), arguments);
+            QObject::connect(runner->process(),
+                QOverload<int, QProcess::ExitStatus>::of(&QProcess::finished), context,
+                    [=, runner = runner.get()](int exitCode,
+                                               QProcess::ExitStatus exitStatus) mutable {
+                state = VirtualMachinePrivate::Existing;
+                if ((exitStatus != QProcess::NormalExit) || (exitCode != 0)) {
+                    functor(state, false);
+                    return;
+                }
+
+                QStringList runningContainers =
+                    QString::fromLocal8Bit(runner->process()
+                        ->readAllStandardOutput()).split('\n', QString::SkipEmptyParts);
+                for (const QString runningContainer : qAsConst(runningContainers)) {
```

Comment:
Better declare `runningContainers` as const. Use `const QString &` for looping.

I would consider breaking the above line before `.split` instead.

Reply:
fixed